### ISSUE TITLE
collapsible_if false positive with RefCell borrow

### ISSUE THREAD

**Issue by awused** [rust-lang/rust-clippy#7513]
Lint name: collapsible_if

I tried this code:

```rust
use std::cell::RefCell;

struct Test {
    a: usize,
    b: RefCell<usize>,
}

fn main() {
    let mut t = Test {
        a: 1,
        b: RefCell::new(2),
    };
                                   
    if *t.b.borrow() == 0 {    
        if (&mut t).a == 5 {       
            println!("a");         
        }                          
    }                              
}                                  
```

Clippy suggests this change which does not compile because it requires borrowing t as mutable while it is already borrowed as immutable.
```rust
 if *t.b.borrow() == 0 && (&mut t).a == 5 {
     println!("a");
 }
 ```

### Meta

- `cargo clippy -V`: clippy 0.1.54 (a178d03 2021-07-26)
- `rustc -Vv`:
  ```
rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-unknown-linux-gnu
release: 1.54.0
LLVM version: 12.0.1
  ```

**Comment by Dilshod070** [comment]
The same problem for me.
@rustbot label +I-suggestion-causes-error

Here is my example of code. clippy suggests second variant, which panics with
`panicked at 'already borrowed: BorrowMutError'`

I'm nоt actually sure if this is bug in clippy or in rust compiler.

```rust
use std::{cell::RefCell, rc::Rc};

struct SomeStruct(u8);

impl SomeStruct {
    pub fn predicate(&self, val: u8) -> bool {
        self.0 == val
    }

    pub fn predicate_mut(&mut self, val: u8) -> bool {
        let res = self.0 == val;
        self.0 += 1;
        res
    }
}

fn main() {
    let rc = Rc::new(RefCell::new(SomeStruct(0)));

    if rc.borrow().predicate(0) {
        if rc.borrow_mut().predicate_mut(0) {
            println!("Everything is OK!")
        }
    }

    if rc.borrow().predicate(1) && rc.borrow_mut().predicate_mut(1) {
        println!("This won't print, should panic")
    }
}
```

**Comment by profetia** [comment]
These two cases are no longer reproducible in the latest nightly. Perhaps it is solved as rustc changed the lifetime behaviors. Anyway this issue can be closed now.
